---
title: "BIOM549 Sequencing Workshop"
subtitle: "Get the updated workshop sheet here: https://ash-bell.github.io/BIOM549-RNA-Seq-Tutorial/"
author: "Ashley Bell"
date: "31/10/2022"
output: rmdformats::material
editor_options: 
  markdown: 
    wrap: 72
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:/Users/agb214/OneDrive - University of Exeter/PhD/BIOM549/")
```

# Introduction
Welcome to the BIOM549 bioinformatics workshop. The goal of this
workshop is to provide you with some basic insight in how RNA-Seq data
sets are analysed. Using this worksheet we will cover the following
Intended Learning Outcomes (ILOs) for this module.

4.  Interpret experimental data obtained using experimental techniques
5.  With minimal guidance, deploy established techniques of analysis,
    practical investigation and enquiry within the biosciences
6.  Critically evaluate experimental data

## Timetable 2022
There are two sessions for this workshop both taught in person. It is
not compulsory to attend but, these are the only times I am teaching. I
would highly recommend you attend them.
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
|Date & Time                  | Location          | Items                               |
|-----------------------------|-------------------|-------------------------------------|
|Wed 30 Nov 13:30 - 16:30     |Old Library 137    |Installing R and analysis walkthrough|
|Tues 13 Dec 09:30 - 12:30    |Old Library 137    |How to present your data             |
"
cat(tabl)
```

### Additional help sessions
I will also make myself available over teams and in person in between
seminars. These are not compulsory but do contact me via teams/email/in-person if you have any issues and I will help you solve them.
```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
|Date & Time             | Location            | Items                        |
|------------------------|---------------------|------------------------------|
|Wed 07 Dec 11:30 - 14:30|Teams / TBC          |Drop in session               |
|Wed 21 Dec 13:30 - 16:30|Teams / TBC          |Drop in session               |
"
cat(tabl)
```
If you have any questions or queries outside of these times, you can
email me at [agb214\@exeter.ac.uk](mailto:agb214@exeter.ac.uk){.email}
and I will get back to you as soon as I can.

## Preface
In your first stream you should have either worked with zebrafish,
plants or fungi. Each of those organisms would have been exposed to
differing environmental conditions, which in turn, may result in them
expressing a different genes in response. Knowledge of which genes are
expressed in different exposure conditions help us understand the host's
response to stimuli. Editing or monitoring these pathways can help increase organism fitness and act as bio-markers of disease. Measuring the difference in gene expression is the goal of RNA sequencing and is what we will be aiming to achieve in this
tutorial. 

## Experimental design
Here you are provided with a data set from the plants stream where  *Arabidopsis thaliana* *pif4* mutants are grown at 22°C and 27°C with samples
collected at 2 and 7 days exposure. Collected samples have their RNA extracted, sequenced and mapped back to their original genome to determine which genes are enriched. From there we get a "count" data set, representing the number of RNA sequences mapping back to each *Arabidopsis* gene. From here, we can devise several hypothesis. 

### Hypothesis
1.    What genes are significantly up-regulated and down-regulated in the *pif4* mutant compared to our controls?
2.    What clusters of genes are differentially express between our exposure groups?
3.    What are the functions of our differentially expressed genes?
4.    What can we concluded about the fitness of our mutant when exposed to higher temperatures?

### Data set link
You can get access to the data sets required for this workshop on the SharePoint link in the "Data" tab. You will need the **count_data.csv** and **metadata.csv** files. 
<https://vle.exeter.ac.uk/course/view.php?id=13271&section=10>

These files will be available on Wed 30 Nov.

# Installation
When you turn up to the sequencing practical, you need to bring your
laptop. On your laptop you need to have R and RStudio installed before
you can analyse any data. You can find instructions on
how to do this below. If you have any issues, please contact me and I'll help you get it installed. 

1.  Download and install R version 4.2.1 for Windows/Mac/Linux from the official R website.This it the programming language we will be using to analyse our data. 
    <https://cran.rstudio.com/> ![](install_R.png)
2.  Install RStudio Desktop (Free version) for Windows/Mac/Linux. This is program that allows us to write in R easily. 
    <https://www.rstudio.com/products/rstudio/download/>
    ![](install_rstudio.png)
3.  After installing R and RStudio, open up RStudio. We next need to install R packages - additional programs written in R to analyse our data. To install R packages, copy the following and paste it into the console window. Press enter to run the code in your console. 
```{r, eval = FALSE}
install.packages(c("tidyverse"))
```
![](install_packages_RStudio.png) 
4. After tidyverse has been installed (it will take a couple minutes), install DESeq2 the same way. Copy and paste the whole code chunk into the console and press enter. This will take a couple of minutes. After this, you're done. You never need to re-install any package you have previously installed. 

```{r, eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c("DESeq2", "apeglm"))
```

# Analysis
This portion starts the analysis. You should have access to two files on
the SharePoint drive accessed through the BIOM549 VLE page. These are
labelled **"metadata.csv"** and **"count_data.csv"**. The metadata table is a
data frame containing which samples belong to which experimental group.
The counts data frame is the output from the pre-processing analysis. This
contains the number of reads that map to each *Arabidopsis* gene. Lets
take a look at the data.

## Set-up
First we need create a script. This is like making a word document and
records all the work you do. To make a script click File --> New File
--> R Script. You can write and run anything here. Click on the save
button and save this file somewhere sensible on your computer. Now
download the metadata and counts file from SharePoint and save it in the
same folder as your script. Everything we are going to do from now on
will be done in this folder.

![](R_screen.png) 
### Setting a working directory
Next we need to tell the computer where is folder is located and
everything we do should be done here. To do that go to Session --> Set
Working Directory --> To Source File Location. A command will now run in
the console that sets your working directory `setwd()`. Copy and paste
this command into your RScript, so that when you run this script again,
it always starts from the same place. Your script should look something
like this.

```{r}
# Copy and paste your working directory here. This is mine and different to yours
setwd("C:/Users/agb214/OneDrive - University of Exeter/PhD/BIOM549/")
```
### Loading in our libraries
If you recall we installed some R libraries previously. Although we never need to re-install our libraries, we need to tell R we want to use them each time we run our script (We never need to re-install Microsoft Word but we do need to tell windows to open it to use it). To do this, we load in our libraries one by one. Its good practice to include all your libraries at the start of your script so anyone else running your script knows
what libraries they need to install to used your code. Copy and paste the following into your script. To run from your script, either click on the line you want to run and press Ctrl+Enter. This will **only** run the line that your cursor is on. To run multiple lines, highlight them all and press Ctrl+Enter. Do this now. 
```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(tidyverse)
```
### Reading in our data sets
Now we can finally load in and look at our data specifying the first column of our data set is our row names. Again copy this into your script and run the code below.
```{r, message=FALSE}
count_data <- read.csv("count_data.csv", row.names = 1)
metadata <- read.csv("metadata.csv", row.names = 1)
```
```{r}
head(count_data)
head(metadata)
```
The command `head()` allows us to look at the first 5 rows of our data.
If we want to take a closer look we can instead use the `View()` command
which will open the data set in a new tab. Beware using `View()` on very
large data sets, you'll freeze your computer! For this exercise, you
should be fine.
```{r, eval = FALSE}
View(count_data)
View(metadata)
```
As you can see, our count data set is the number of reads that mapped to each of the *Arabidopsis* genes for each of our samples. The metadata data set tells us which sample belongs to which exposure group. Pretty simple. Now lets see if any of the genes are significantly enriched (or not!) in our data set compare to the controls. 

## Analysis
Next it is important the column names from our count data in the same
order as the row names for the metadata. Without this, our downstream
software will combine the data incorrectly. Both these commands should
return "TRUE".

```{r}
# Are all the column names in our count data set in the row names of our metadata?
all(colnames(count_data) %in% rownames(metadata))
# Are the column names in our count data set in the exact sample order as the row names of our metadata?
all(colnames(count_data) == rownames(metadata))
```
Notice how if we put a "#" in front of the a line of code, it doesn't run. A "#" tells R, don't run this code, it is a comment. Comments are very important to tell other people what a line of code does. As codes get more complicated, other people may want to modify your code for their purposes. Without a comment this if very hard to do. I recomend you do the same for your own notes. 

### Combining data sets into a DESeq2 object
Next we need to combine our count data with our metadata. We use a
package called DESeq2 which helps us keep track of multiple data sets in
one experiment. Before we do that, we need to tell DESeq2 which is our
control and experimental groups. To do this we set the column
"Experimental group" as a factor, meaning they are categories and the
first level (category) is "wildtype" or our control. Now we're good to
go.
```{r}
metadata$Experimental_group <- as.factor(metadata$Experimental_group)
metadata$Experimental_group <- relevel(metadata$Experimental_group, ref = "wildtype")

dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = metadata,
                              design = ~Experimental_group)

dds
```
Notice when we type in `dds` into our console, we now have a
DESeqDataSet. It has told us we have a metadata table and count assay, with
our row names as genes, columns as samples and metadata categories as
Temperature, Experimental_group, Replicate and Days_exposure. Great!

### Pre-filtering
Next we should remove genes that aren't expressed frequently. This
removes the "noise" which helps us find significant trends. These genes
that are rarely expressed need to be in higher quantities so we can be
sure they are significantly expressed, so we remove them. The alternatives is to use [shrinkage methods](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#alternative-shrinkage-estimators), which we will not over here. Therefore, if the gene is not expressed over 1000 times in our entire data set, it is
removed. 1000 is an arbitrary number and could be change depending on
the study, for example you could remove everything below the 95th
percentile.
```{r}
keep <- rowSums(counts(dds)) >= 1000
dds <- dds[keep, ]
```

Another example is remove any gene count that isn't at least 1000 in X
samples.
```{r}
# Get the length of the number of columns in our "count_data" data set
X <- length(colnames(count_data))
# Subset our DESeq2 object where the minimum gene count across our entire data set is over 1000 per gene
keep <- rowSums(counts(dds) >= 1000) >= X
dds <- dds[keep, ]
```

It is up to you to decide what filter to use. For this tutorial I will
remove any gene count that is not at least 1000 in all samples. The
great thing about DDSeq2 is it does all our calculations for us. There
are
```{r, echo=F}
print(nrow(count_data))
```
genes in our data set. Imagine manually filtering each one out! This is
the power of using a coding language to analyse large data sets.

## Differential expression analysis
Now we can calculate if the genes are statistically differential expressed.
We use the `p.adjusted` value to account for multiple retesting - the idea
that if you test for everything, by random chance you will find
something that is significant but its just a "lucky" pattern. If your interested, you can find out more [here](https://physiology.med.cornell.edu/people/banfelder/qbio/resources_2008/1.5_GenespringMTC.pdf).

```{r, message=FALSE}
dds <- DESeq(dds)

# Define our p-value (alpha) cutoff as 0.05
res <- results(dds, alpha = 0.05)

# Give us a quick summary of the results
# LFC > 0 (up) is number of up regulated genes
# LFC < 0 (down) is the number of down regulated genes
summary(res)
```
```{r, echo = FALSE}
print(paste0("Number of significantly up/down regulated genes: ", sum(res$padj < 0.05, na.rm=TRUE)))
print(paste0("Total number of genes in this dataset: ", nrow(assay(dds))))
```

### Plotting
So now we know there are a total number of genes that are significantly
up/down regulated. But how does this look compared to the entire data set? We can plot this as a volcano plot.
```{r}
plotMA(res, xlim=c(1000,20000), ylim=c(-0.5,0.5))
```
The dots in blue are significantly up/down regulated and the ones in
grey not. Lets take a closer look at the most significantly different
regulated gene.

```{r}
plotCounts(dds, gene=which.min(res$padj), intgroup="Experimental_group")
```
We can definitely see this gene is down-regulated in the *pif4* mutant

## Multi-factor analysis
So we have concluded there are genes that are differentially expressed in our *pif4* mutant compare to our control. However, we have not taken into account that our experiment has multiple parameters, not just two different experimental groups (*pif4* and wildtype). DESeq2 can account for this if you tell it. Lets do
that and see the results. Here we define all our groups, Temperature,
Experimental group (wildtype or mutant) and Days exposure (number of
days exposed to each temperature).Here we also include an interaction
term: `Days_exposure:Temperature` which tells us if the number of days
exposed to each temperature together have an compunded effect compared to if they
were individually measured.
```{r}
ddsMF <- dds
design(ddsMF) <- formula(~ Temperature + Experimental_group + Days_exposure, Days_exposure:Temperature)
ddsMF <- DESeq(ddsMF)
resMF <- results(ddsMF, alpha = 0.05)
summary(resMF)
```

Over half of the genes are up-/down-regulated. However, its expected
that at different temperatures and times gene expression will be
different. Lets just focus if these variables still have an effect on
the wildtype versus *pif4* mutant
```{r}
resMF <- results(ddsMF, contrast = c("Experimental_group", "wildtype", "pif4_mutant"), alpha = 0.05)
summary(resMF)
```
It still looks like they do. Great! Now lets look at the samples as a whole in a bit more detail. 

## Count transformations
When working with differential expression, we use raw counts as we are
treating each sample discretely. However, if we want to cluster groups (combined multiple samples), we should transform the raw counts to be relative to each other. DESeq2 has a number of transformation functions, such as `vst`, `rlog` and `normTransform`. There are
advantages and disadvantages if each which I won't cover, but you can find out more [here](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#how-do-i-use-vst-or-rlog-data-for-differential-testing). Instead,
I'm going to use `vst`, short for Variance Stabilizing Transformation and `blind = FALSE` as I want it to be aware we have different exposure groups.
```{r}
vsd <- vst(ddsMF, blind=FALSE)
```
Our data set has now been normalised so we can see how different each group clusters

## Replicate clustering
First we need to see how closely our replicates cluster. Are they truly different compared to the controls or is everything random. But how do we visualise 1000s of genes in one plot? We use a method called dimension reduction techniques, which condenses a data set into a single point. There are many methods to do this, but in this case we will use a Principle Component Analysis (PCA). Here I have plotted the two PCA plots, one coloured by experimental group, the other by days exposure and temperature. 
```{r}
plotPCA(vsd, intgroup = c("Experimental_group"))

plotPCA(vsd, intgroup = c("Days_exposure", "Temperature"))
```
Our replicates cluster based on the exposure conditions, showing us that there are differences in gene expression across conditions but not within replicates (Good!). 

# WORK IN PROGRESS

#https://hbctraining.github.io/DGE_workshop_salmon_online/lessons/08a_DGE_LRT_results.html

##############################
#Allow for more space around the borders of the plot
par(mar = c(5, 4, 4, 4))

#Set your log-fold-change and p-value thresholds
lfc = 0.5
pval = 0.05

LFC <- resMF
#make a data frame with the log2 fold-changes and adjusted p-values
tab = data.frame(logFC = LFC$log2FoldChange, negLogPval = -log10(LFC$padj))

plot(tab, pch = 16, cex = 0.4, xlab = expression(log[2]~fold~change),
     ylab = expression(-log[10]~pvalue)) #replace main = with your title

#Genes with a fold-change greater than 2 and p-value<0.05:
signGenes = (abs(tab$logFC) > lfc & tab$negLogPval > -log10(pval))

#Colour these red
points(tab[signGenes, ], pch = 16, cex = 0.5, col = "red")

#Show the cut-off lines
abline(h = -log10(pval), col = "green3", lty = 2)
abline(v = c(-lfc, lfc), col = "blue", lty = 2)

mtext(paste("FDR =", pval), side = 4, at = -log10(pval), cex = 0.6, line = 0.5, las = 1)
mtext(c(paste("-", lfc, "fold"), paste("+", lfc, "fold")), side = 3, at = c(-lfc, lfc),
      cex = 0.6, line = 0.5)
      
###########################

#Select your chosen gene 
d <- plotCounts(dds, gene=which.min(res$padj), 
                intgroup=c("Experimental_group", "Temperature"), 
                returnData = T)

ggplot(d, aes(x = Experimental_group, y = count, fill = as.factor(Temperature))) + 
  geom_boxplot()  + 
  geom_dotplot(binaxis = 'y', stackdir = 'center')

print(p)

####################




resSig <- subset(resMF, padj < 0.05)

ggplot(d, aes(x=Experimental_group, y=count)) +
geom_point(position=position_jitter(w=0.1,h=0)) +
scale_y_log10(breaks=c(25,100,400))

ddsMF <- DESeq(ddsMF)

resMF <- results(ddsMF, alpha = 0.05)

vsd <- vst(dds, blind=FALSE) 
plotPCA(vsd, intgroup = c("Days_exposure",
"Temperature"))

pcaData <- plotPCA(vsd, intgroup=c("Days_exposure", "Temperature"),
returnData=TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar")) ggplot(pcaData,
aes(PC1, PC2, color=as.factor(Temperature),
shape=as.factor(Days_exposure))) + geom_point(size=3) +
xlab(paste0("PC1: ",percentVar\[1\],"% variance")) + ylab(paste0("PC2:
",percentVar\[2\],"% variance")) + coord_fixed()
